macros/provider.go

```diff
@@ -0,0 +1,155 @@
+	macroProvider := &macroProvider{macros: map[string]string{}}
+	macroProvider.populateRequestMacros(reqWrapper)
+	return macroProvider
+}
+
+func (b *macroProvider) populateRequestMacros(reqWrapper *openrtb_ext.RequestWrapper) {
+	b.macros[MacroKeyTimestamp] = strconv.Itoa(int(time.Now().Unix()))
+	reqExt, err := reqWrapper.GetRequestExt()
+	if err == nil && reqExt != nil && reqExt.GetPrebid() != nil {
+		for key, value := range reqExt.GetPrebid().Macros {
+			customMacroKey := CustomMacroPrefix + key       // Adding prefix PBS-MACRO to custom macro keys
+			b.macros[customMacroKey] = truncate(value, 100) // limit the custom macro value  to 100 chars only
+		}
+
+		if reqExt.GetPrebid().Integration != "" {
+			b.macros[MacroKeyIntegration] = reqExt.GetPrebid().Integration
+		}
+		channel := reqExt.GetPrebid().Channel
+		if channel != nil {
+			b.macros[MacroKeyChannel] = channel.Name
+		}
+
+	}
+	b.macros[MacroKeyAuctionID] = reqWrapper.ID
+	if reqWrapper.App != nil {
+		if reqWrapper.App.Bundle != "" {
+			b.macros[MacroKeyAppBundle] = reqWrapper.App.Bundle
+		}
+
+		if reqWrapper.App.Domain != "" {
+			b.macros[MacroKeyDomain] = reqWrapper.App.Domain
+		}
+
+		if reqWrapper.App.Publisher != nil {
+			if reqWrapper.App.Publisher.Domain != "" {
+				b.macros[MacroKeyPubDomain] = reqWrapper.App.Publisher.Domain
+			}
+			if reqWrapper.App.Publisher.ID != "" {
+				b.macros[MacroKeyAccountID] = reqWrapper.App.Publisher.ID
+			}
+		}
+	}
+
+	if reqWrapper.Site != nil {
+		if reqWrapper.Site.Page != "" {
+			b.macros[MacroKeyPageURL] = reqWrapper.Site.Page
+		}
+
+		if reqWrapper.Site.Domain != "" {
+			b.macros[MacroKeyDomain] = reqWrapper.Site.Domain
+		}
+
+		if reqWrapper.Site.Publisher != nil {
+			if reqWrapper.Site.Publisher.Domain != "" {
+				b.macros[MacroKeyPubDomain] = reqWrapper.Site.Publisher.Domain
+			}
+
+			if reqWrapper.Site.Publisher.ID != "" {
+				b.macros[MacroKeyAccountID] = reqWrapper.Site.Publisher.ID
+			}
+		}
+	}
+
+	userExt, err := reqWrapper.GetUserExt()
+	if err == nil && userExt != nil && userExt.GetConsent() != nil {
+		b.macros[MacroKeyConsent] = *userExt.GetConsent()
+	}
+	if reqWrapper.Device != nil && reqWrapper.Device.Lmt != nil {
+		b.macros[MacroKeyLmtTracking] = strconv.Itoa(int(*reqWrapper.Device.Lmt))
+	}
+
+}
+
+func (b *macroProvider) GetMacro(key string) string {
+	return url.QueryEscape(b.macros[key])
+}
+
+func (b *macroProvider) PopulateBidMacros(bid *entities.PbsOrtbBid, seat string) {
+
+	if bid.GeneratedBidID != "" {
+		b.macros[MacroKeyBidID] = bid.GeneratedBidID
+	} else {
+		b.macros[MacroKeyBidID] = bid.Bid.ID
+	}
+	b.macros[MacroKeyBidder] = seat
+}
+
+func (b *macroProvider) PopulateEventMacros(vastCreativeID, eventElement, vastEventType string) {
+	b.macros[MacroKeyVastCRTID] = vastCreativeID
+	b.macros[MacroKeyVastEvent] = vastEventType
+	b.macros[MacroKeyEventType] = eventElement
+}
+
+func truncate(text string, width uint) string {
```

Comment:
Could you please write some test cases that covers this function as well? Perhaps a macro value larger than the 100 char limit?

Reply:
Added unit test